### PR TITLE
Add support for analysing aggregate handlers.

### PR DESCRIPTION
**Pull request by danilvpetrov**
#### What change does this introduce?

This change is the next step to introduce the static analysis of the configuration data of Dogma applications. This particular change concentrates on analysing Dogma application aggregate handlers.

#### Why make this change?

This change is required as the next step to implement the feature of loading Dogma application configurations using static analysis.

#### Is there anything you are unsure about?

The aggregate handler tests are using gomega's `gstruct` package. They do make the test look bulkier and I am not sure if it's a right fit.

#### What issues does this relate to?

Partially addresses #104.

### CODE REVIEW

File: static/analysis.go [hunk 1]
```diff
@@ -69,3 +80,63 @@ func analyzeIdentityCall(c *ssa.Call) configkit.Identity {
 
 	return ident
 }
+
+// analyzeHandler analyzes the application's handler from the given type to
+// deduce its handler configuration.
+//
+// It is assumed that the caller knows what type this handler is and passes
+// handler's type through ht parameter.
+func analyzeHandler(
```
Comment (jmalloc) [inline static/analysis.go:89]:
```suggestion
// analyzeHandler analyzes a type that implements one of the Dogma handler
// interfaces to deduce its handler configuration.
func analyzeHandler(
```

File: static/analysis.go [hunk 2]
```diff
@@ -1,32 +1,43 @@
...
 
 	for _, c := range findConfigurerCalls(fn) {
+		args := c.Common().Args
+
 		switch c.Common().Method.Name() {
 		case "Identity":
 			app.IdentityValue = analyzeIdentityCall(c)
+		case "RegisterAggregate":
+			app.HandlersValue.Add(
+				analyzeHandler(
```
Comment (jmalloc) [inline static/analysis.go:35]: What happens if the type passed to RegisterAggregate doesn't implement `dogma.AggregateMessageHandler`, I guess compilation would have failed and resulted in a `nil` package earlier?

File: static/analysis.go [hunk 3]
```diff
@@ -1,32 +1,43 @@
...
+					prog,
+					args[0].(*ssa.MakeInterface).X.Type(),
```
Comment (jmalloc) [inline static/analysis.go:37]: Can you always rely on the argument being one of these `MakeInterface` types? What if, for example, the argument was a variable of type `dogma.AggregateMessageHandler` ?

File: static/analysis.go [hunk 4]
```diff
@@ -69,3 +80,63 @@ func analyzeIdentityCall(c *ssa.Call) configkit.Identity {
...
+	fn := prog.LookupMethod(typ, pkg, "Configure")
+
+	for _, c := range findConfigurerCalls(fn) {
+		args := c.Common().Args
+
+		switch c.Common().Method.Name() {
+		case "Identity":
+			hdr.IdentityValue = analyzeIdentityCall(c)
+		case "ConsumesCommandType":
+			hdr.MessageNamesValue.Produced.Add(
```
Comment (jmalloc) [inline static/analysis.go:113]:
```suggestion
			hdr.MessageNamesValue.Consumed.Add(
```

File: static/analysis.go [hunk 5]
```diff
@@ -69,3 +80,63 @@ func analyzeIdentityCall(c *ssa.Call) configkit.Identity {
...
+				message.NameFromType(
+					args[0].(*ssa.MakeInterface).X.Type(),
+				),
+				message.CommandRole,
+			)
+		case "ProducesEventType":
+			hdr.MessageNamesValue.Consumed.Add(
```
Comment (jmalloc) [inline static/analysis.go:120]:
```suggestion
			hdr.MessageNamesValue.Produced.Add(
```

File: static/analysis.go [hunk 6]
```diff
@@ -69,3 +80,63 @@ func analyzeIdentityCall(c *ssa.Call) configkit.Identity {
...
+				message.EventRole,
+			)
+		}
+	}
+
+	return hdr
+}
+
+// getTypePkg returns a package from the given type.
+func getTypePkg(typ types.Type) *types.Package {
```
Comment (jmalloc) [inline static/analysis.go:133]:
```suggestion
// pkgOfNamedType returns the package in which typ is declared.
//
// It panics if typ is not a named type or a pointer to a named type.
func pkgOfNamedType(typ types.Type) *types.Package {
```

File: static/analysis.go [hunk 7]
```diff
@@ -69,3 +80,63 @@ func analyzeIdentityCall(c *ssa.Call) configkit.Identity {
...
+	switch t := typ.(type) {
+	case *types.Named:
+		return t.Obj().Pkg()
+	case *types.Pointer:
+		return t.Elem().(*types.Named).Obj().Pkg()
```
Comment (jmalloc) [inline static/analysis.go:138]: If this isn't a named type it will panic with a type assertion failure rather than the meaningful panic message you wrote below.

```suggestion
		return pkgOfNamedType(t.Elem())
```

File: static/analysis.go [hunk 8]
```diff
@@ -69,3 +80,63 @@ func analyzeIdentityCall(c *ssa.Call) configkit.Identity {
...
+	default:
+		panic(fmt.Sprintf("cannot analyze package for type %v", typ))
```
Comment (jmalloc) [inline static/analysis.go:140]:
```suggestion
		panic(fmt.Sprintf("cannot determine package for anonymous or built-in type %v", typ))
```

File: static/analysis.go (hunk 3 above)
Comment (jmalloc) [reply static/analysis.go:37]: Or `nil`, for that matter. Obviously this isn't normal but it would still compile and we'd want to make sure that this package didn't choke on that input.

**Review by jmalloc — CHANGES_REQUESTED** [review]

File: static/analysis.go (hunk 2 above)
Comment (danilvpetrov) [reply static/analysis.go:35]: Correct, this will be a typical package error that would result in `nil` package as an output from the `ssa.LoadPackages()` function.

File: static/analysis.go (hunk 3 above)
Comment (danilvpetrov) [reply static/analysis.go:37]: Good point, I will write the code and tests to cover for this scenarios.

File: static/analysis.go (hunk 2 above)
Comment (jmalloc) [reply static/analysis.go:35]: Ok, cool!

File: static/analysis.go (hunk 3 above)
Comment (danilvpetrov) [reply static/analysis.go:37]: Ok, even if we assign an instance of aggregate to a variable of type `dogma.AggregateMessageHandler` and pass it as an argument to the method, `ssa` packages still interprets that as `*ssa.MakeInterface` instruction.

However, passing `nil` is treated as `*ssa.Const` which causes the code above to explode. 

Should we do a check similar to this and write tests to address this scenario?

```go
            case "RegisterAggregate":
			if mi, ok := args[0].(*ssa.MakeInterface); ok {
				app.HandlersValue.Add(
					analyzeHandler(
						prog,
						mi.X.Type(),
						configkit.AggregateHandlerType,
					),
				)
			}
		}
```

Comment (jmalloc) [reply static/analysis.go:37]: Can you try what happens if there is a typed `const` that is a real implementation of `AggregateMessageHandler`?

like ...

```go
type SomeHandler string

const Handler SomeHandler = "..."

func (Blah) Configure(........
```

Because I wonder if that is a `*ssa.Const` or not, since it obvious IS a const, but also implements the interface.

Comment (danilvpetrov) [reply static/analysis.go:37]: Funny thing that the type above still ends up as `*ssa.MakeInterface`: 
```golang
&ssa.MakeInterface{
    register: ssa.register{
        anInstruction: ssa.anInstruction{
			block:(*ssa.BasicBlock)(0xc00364c790),
		},
		num:3,
		typ:(*types.Named)(0xc001200140),
		pos:0,
		referrers:[]ssa.Instruction{
			(*ssa.Call)(0xc003f92280),
		},
	},
	X:(*ssa.UnOp)(0xc005177500),
}
```

Comment (danilvpetrov) [reply static/analysis.go:37]: 🤷

Comment (jmalloc) [reply static/analysis.go:37]: Ok, cool.. I guess it represents the "boxing" of the type into the underlying interface struct?

I think the `ok` style type assertion is the way to go then.

Comment (danilvpetrov) [reply static/analysis.go:37]: Yep, boxing would be exactly what's happening here in my mind too. I will still throw some tests for passing `nil` in `RegisterAggregate()`, `ConsumesCommandType()`, and `ProducesEventType()` then.

Comment (danilvpetrov) [reply static/analysis.go]: Ok, added type assertions and tests in https://github.com/dogmatiq/configkit/pull/123/commits/6c01fa7d4e0c70a83b8c062dfd27e415c23177fc.

File: static/testdata/aggregates/single-aggregate-app/aggregate.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+func (AggregateHandler) Configure(c dogma.AggregateConfigurer) {
+	c.Identity("<aggregate>", "ef16c9d1-d7b6-4c99-a0e7-a59218e544fc")
+
+	c.ConsumesCommandType(fixtures.MessageA{})
+	c.ConsumesCommandType(fixtures.MessageB{})
+	c.ConsumesCommandType(fixtures.MessageC{})
+
+	c.ProducesEventType(fixtures.MessageD{})
+	c.ProducesEventType(fixtures.MessageE{})
+	c.ProducesEventType(fixtures.MessageF{})
```
Comment (jmalloc) [inline static/testdata/aggregates/single-aggregate-app/aggregate.go:34]: In the interest of reducing the size and complexity of the tests, you could reduce the number of message types used here to just 2 of each instead of 3.

Likewise, since this test covers the case where there are multiple message types the other tests don't need to duplicate that; they could just have a single command and a single event.

I am only suggesting this to reduce clutter, no big deal if you want to retain them.

File: static/aggregate_test.go [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+			}
+
+			pkgs, err := packages.Load(&cfg, "./...")
+			Expect(err).NotTo(HaveOccurred())
+
+			apps := FromPackages(pkgs)
+
+			Expect(apps).To(
+				ConsistOf(
+					PointTo(
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply static/aggregate_test.go:30]: These tests have made me aware of how dependent they are on the internal fields of the entity structs. It might be really tricky to update these tests if the implementation of those entity structs were to change. 

I think we should spend some time to workout how we can test this via the `configkit` interfaces, without expecting a specific implementation. 

In addition to that I think it would be good to focus the expectations on the handler configs, and less so on the application in an effort to reduce the cognitive overhead of working out what went wrong when one of the expectations does fail.

**Review by jmalloc — CHANGES_REQUESTED** [review]
I think this looks really good, this kind of code is very often indecipherable but I think this has really readable :+1:

My other comments on the PR are all concerned with the complexity / intricacy of the tests. We need to develop some nicer patterns for testing these deeply nested structures as it's only going to explode as the additional handlers are introduced.

File: static/analysis.go (hunk 3 above)
Comment (jmalloc) [reply static/analysis.go]: :+1:

File: static/testdata/aggregates/single-aggregate-app/aggregate.go (hunk 9 above)
Comment (danilvpetrov) [reply static/testdata/aggregates/single-aggregate-app/aggregate.go]: Done, the number of messages reduced in https://github.com/dogmatiq/configkit/pull/123/commits/f5ca4975c9c602e436921b0e95a36264bab66712.

File: static/aggregate_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (danilvpetrov) [reply static/aggregate_test.go]: Simplified the tests as per the suggestion above in https://github.com/dogmatiq/configkit/pull/123/commits/2d1ce67c9a4b0f59e72e72a333aaccf30496e10f.  The first single aggregate handler test still uses `github.com/onsi/gomega/gstruct` assertions because of multiple messages for each type in `message.NameRoles` maps. All remaining tests are now using direct `Equal()` assertions.

[the remark this answers is not in the document]
Comment (danilvpetrov) [reply static/aggregate_test.go]: Another go in https://github.com/dogmatiq/configkit/pull/123/commits/23b4fd9461639f5459c139229b0f847640b4af82 on simplifying the aggregate tests as we agreed during our recent talk. This time aI am making sure that we are not using any internal types for assertions.

File: static/aggregate_test.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+			}
+
+			pkgs, err := packages.Load(&cfg, "./...")
+			Expect(err).NotTo(HaveOccurred())
+
+			apps := FromPackages(pkgs)
+			Expect(apps).To(HaveLen(1))
+			Expect(apps[0].Handlers().Aggregates()).To(HaveLen(2))
+
+			a1 := apps[0].Handlers().Aggregates()[0]
```
Comment (danilvpetrov) [inline static/aggregate_test.go:76]: This is relying on the fact that `ssautil.AllPackages()` function parses files within a package in  alphabetic order . To me, there is a concern that this can potentially make the tests fragile in case the logic within `ssautil.AllPackages()` changes.

Comment (danilvpetrov) [reply static/aggregate_test.go]: To cope with this problem I am proposing the approach highlighted in https://github.com/dogmatiq/configkit/pull/123/commits/9614e85ff4777a5414e9b4dbf5c64555662f320d.  IMHO, using `configkit.HandlerSet.ByIdentity()` simplifies the aggregate retrieval and yet we still test its remaining components regardless of the aggregate order in `configkit.HandlerSet.Aggregates()`.

Comment (danilvpetrov) [reply static/aggregate_test.go]: A slice of aggregate identities is now used to test multiple aggregate handlers in https://github.com/dogmatiq/configkit/pull/123/commits/b2bd8a448a89729750a8c1eaba730119f297f439.

File: static/aggregate_test.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+			pkgs, err := packages.Load(&cfg, "./...")
+			Expect(err).NotTo(HaveOccurred())
+
+			apps := FromPackages(pkgs)
+			Expect(apps).To(HaveLen(1))
+			Expect(apps[0].Handlers()).To(Equal(configkit.HandlerSet{}))
+		})
+	})
+
+	When("a nil value passed as a message while configuring an aggregate handler", func() {
```
Comment (jmalloc) [inline static/aggregate_test.go:109]:
```suggestion
	When("a nil value is passed as a message", func() {
```

Does this test cover both commands and events?

File: static/aggregate_test.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+		It("skips the nil value message", func() {
```
Comment (jmalloc) [inline static/aggregate_test.go:110]:
```suggestion
		It("does not add the message to the aggregate configuration", func() {
```

**Review by jmalloc — CHANGES_REQUESTED** [review]

File: static/aggregate_test.go (hunk 12 above)
Comment (jmalloc) [reply static/aggregate_test.go]: Do we need a separate test here for commands vs events?

Comment (danilvpetrov) [reply static/aggregate_test.go]: It does cover both commands and events at the moment all in one test, I can split it quite easily though.

Comment (jmalloc) [reply static/aggregate_test.go]: Nah, all good.. I think this is good to go man!

**Review by jmalloc — APPROVED** [review]
